From 8b81238787cafa846673058e002a1edfe05da76b Mon Sep 17 00:00:00 2001
From: Wang Quanyang <quanyang.wang@windriver.com>
Date: Thu, 30 Mar 2017 03:12:59 +0000
Subject: [PATCH] ASoC: imx-HiFi-ASRC-FE: set nonatomic flag of
 snd_soc_dai_link structure

The function sdma_free_bd will call dma_free_coherent which must be run in
irq-enable context. So set nonatomic flag of snd_soc_dai_link structure to
avoid snd_pcm_drop disabling irq. Or else it will occur calltrace as below:

------------[ cut here ]------------
WARNING: CPU: 2 PID: 752 at /kernel-source/include/linux/dma-mapping.h:4c
Modules linked in:
CPU: 2 PID: 752 Comm: aplay Not tainted 4.8.18-WR9.0.0.4_standard #2
Hardware name: Freescale i.MX6 Quad/DualLite (Device Tree)
[<8011246c>] (unwind_backtrace) from [<8010c454>] (show_stack+0x20/0x24)
[<8010c454>] (show_stack) from [<804fb5d8>] (dump_stack+0x78/0x94)
[<804fb5d8>] (dump_stack) from [<8012f7bc>] (__warn+0xf0/0x110)
[<8012f7bc>] (__warn) from [<8012f8ac>] (warn_slowpath_null+0x30/0x38)
[<8012f8ac>] (warn_slowpath_null) from [<805857d4>] (sdma_free_bd+0xbc/0x10c)
[<805857d4>] (sdma_free_bd) from [<80585844>] (sdma_desc_free+0x20/0x2c)
[<80585844>] (sdma_desc_free) from [<80584c5c>] (vchan_dma_desc_free_list+0x78/0x7c)
[<80584c5c>] (vchan_dma_desc_free_list) from [<80587d54>] (sdma_terminate_all+0x140/0x1ac)
[<80587d54>] (sdma_terminate_all) from [<808623b4>] (fsl_asrc_dma_trigger+0x1c4/0x1f8)
[<808623b4>] (fsl_asrc_dma_trigger) from [<80854a60>] (soc_pcm_trigger+0xc8/0x124)
[<80854a60>] (soc_pcm_trigger) from [<808552d4>] (dpcm_fe_dai_do_trigger+0x74/0x138)
[<808552d4>] (dpcm_fe_dai_do_trigger) from [<808553dc>] (dpcm_fe_dai_trigger+0x44/0x48)
[<808553dc>] (dpcm_fe_dai_trigger) from [<8083e024>] (snd_pcm_do_stop+0x58/0x60)
[<8083e024>] (snd_pcm_do_stop) from [<8083ddc8>] (snd_pcm_action_single+0x48/0x74)
[<8083ddc8>] (snd_pcm_action_single) from [<8083dea0>] (snd_pcm_action+0x34/0xf4)
[<8083dea0>] (snd_pcm_action) from [<8083df84>] (snd_pcm_stop+0x24/0x2c)
[<8083df84>] (snd_pcm_stop) from [<8083f430>] (snd_pcm_drop+0x78/0x9c)
[<8083f430>] (snd_pcm_drop) from [<8083f484>] (snd_pcm_release_substream+0x30/0xc8)
[<8083f484>] (snd_pcm_release_substream) from [<8083f560>] (snd_pcm_release+0x44/0x90)
[<8083f560>] (snd_pcm_release) from [<8026c6f4>] (__fput+0xe8/0x1c4)
[<8026c6f4>] (__fput) from [<8026c860>] (____fput+0x18/0x1c)
[<8026c860>] (____fput) from [<8014bdcc>] (task_work_run+0xcc/0xe0)
[<8014bdcc>] (task_work_run) from [<80133f28>] (do_exit+0x4c8/0xa14)
[<80133f28>] (do_exit) from [<80134588>] (do_group_exit+0xc8/0xcc)
[<80134588>] (do_group_exit) from [<801345ac>] (__wake_up_parent+0x0/0x30)
[<801345ac>] (__wake_up_parent) from [<801077e0>] (ret_fast_syscall+0x0/0x3c)
---[ end trace 5608bf70c694c0a6 ]---

Signed-off-by: Wang Quanyang <quanyang.wang@windriver.com>
---
 sound/soc/fsl/imx-wm8962.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/sound/soc/fsl/imx-wm8962.c b/sound/soc/fsl/imx-wm8962.c
index fa13865..9b122bb 100644
--- a/sound/soc/fsl/imx-wm8962.c
+++ b/sound/soc/fsl/imx-wm8962.c
@@ -659,6 +659,7 @@ audmux_bypass:
 		data->dai[1].cpu_of_node = asrc_np;
 		data->dai[1].platform_of_node = asrc_np;
 		data->dai[1].dynamic = 1;
+		data->dai[1].nonatomic = 1;
 		data->dai[1].ignore_pmdown_time = 1;
 		data->dai[1].dpcm_playback = 1;
 		data->dai[1].dpcm_capture = 1;
-- 
1.7.5.4

